### PR TITLE
BLD, TST: Build and test Pyodide wheels for `pandas` in CI

### LINKED ISSUE #57891: ENH: out-of-tree Pyodide builds in CI for `pandas`
### Feature Type

- [x] Adding new functionality to pandas
- [ ] Changing existing functionality in pandas
- [ ] Removing existing functionality in pandas

### Problem Description

This feature proposes adding out-of-tree Pyodide builds for `pandas` on its own CI. As opposed to in-tree builds in the Pyodide packages repository (see the [recipe](https://github.com/pyodide/pyodide/blob/main/packages/pandas/meta.yaml)), this will enable building WASM wheels (which are 32-bit)[^1] via [the Emscripten toolchain](https://emscripten.org/) on GitHub Actions.

In my most recent work assignment, I am working on improving the interoperability for the Scientific Python ecosystem of packages with Pyodide and with each other (https://github.com/Quansight-Labs/czi-scientific-python-mgmt/issues/18), which shall culminate with efforts towards bringing interactive documentation for these packages where they can then be run in [JupyterLite notebooks](https://jupyter.org/try-jupyter/lab/index.html), through nightly builds and wheels for these packages pushed to PyPI-like indices on Anaconda.org (https://github.com/Quansight-Labs/czi-scientific-python-mgmt/issues/19) at and during a later phase during the project.

[^1]: It is to be noted that #15889 removed 32-bit support by removing the wheels, but if I understand correctly, `pandas` continues to be supported when building from source on 32-bit platforms.

### Feature Description

1. A CI pipeline via GitHub Actions that builds and tests `pandas` in an activated virtual environment created and handled by Pyodide.
2. Fixing up the tests wherever applicable, conforming to [limitations](https://pyodide.org/en/stable/usage/wasm-constraints.html) that come from the Emscripten toolchain (current limitations include the lack of threading support and for running subprocesses, limited file system access, and others).

### Alternative Solutions

N/A

### Additional Context

I notice that the build and the tests are going to be split into separate packages in #53007, which can be handled (I assume that the imminent `pandas-tests` package will be a pure Python one, which does not need compilation for Pyodide – they are supported by default).

cc @lithomas1 for any additional comments here, as required. Thomas and I have been in conversation about this just about a fortnight ago, and we had agreed that reducing the size of the wheels would be helpful for the WebAssembly rendition of `pandas` as well – it is particularly useful where reducing bandwidth usage is in the question. I am happy to follow along on further developments on that regard.

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
- [x] closes #57891 
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

## Description

This PR adds a workflow, `emscripten.yml`, to the GitHub Actions CI activities where WebAssembly wheels for `pandas` using Pyodide/Emscripten are built and then tested. I have fixed up a few of the tests, but most of them are skipped for now and these differences _probably_ do not affect major functionality.

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
Here is one of the latest workflow runs on my fork: https://github.com/agriyakhetarpal/pandas/actions/runs/8331354944/job/22798065116

The associated failures right now are mostly coming from a singular test that (which is associated with a hypothesis generator that brings additional test cases?)

For example, there is a floating point inaccuracy, here – which could be corrected by setting the relative and absolute tolerances differently:

```
E           AssertionError: DataFrame.iloc[:, 2] (column name="dt_as_dt") are different
E           
E           DataFrame.iloc[:, 2] (column name="dt_as_dt") values are different (100.0 %)
E           [index]: [0, 1]
E           [left]:  [253402127999999, 1564703999999]
E           [right]: [253402127999998, 1564703999998]
```

and some `tzlocal` tests, which seem to bring overflow errors. Note: this output is truncated, please see the entirety of them in the workflow logs:

```
_______________ test_apply_out_of_range[tzlocal()-BusinessHour] ________________
[XPASS(strict)] OverflowError inside tzlocal past 2038
_____________ test_apply_out_of_range[tzlocal()-BusinessMonthEnd] ______________
[XPASS(strict)] OverflowError inside tzlocal past 2038
____________ test_apply_out_of_range[tzlocal()-BusinessMonthBegin] _____________
[XPASS(strict)] OverflowError inside tzlocal past 2038
```

I am not completely sure how to fix these tests – I don't get why these aren't skipped completely? I would appreciate another pair of eyes on this; happy to receive pointers from someone experienced with the codebase :)

**Comment by WillAyd** [comment]
> I am not completely sure how to fix these tests – I don't get why these aren't skipped completely? I would appreciate another pair of eyes on this; happy to receive pointers from someone experienced with the codebase :)

At least for the second test it looks like it is imperatively skipping for platforms that are not 64 bit. I take it the platform these are running on is still 64 bit but we are cross compiling to a 32 bit WASM target?

https://github.com/pandas-dev/pandas/blob/495f80896852e450badc8866ee7ebe8c434fa228/pandas/tests/indexes/datetimes/methods/test_resolution.py#L25

**Comment by agriyakhetarpal** [comment]
> At least for the second test it looks like it is imperatively skipping for platforms that are not 64 bit. I take it the platform these are running on is still 64 bit but we are cross compiling to a 32 bit WASM target?

Hi, @WillAyd, yes – cross-compilation is indeed what is happening here and the platform is 64-bit, but the Python interpreter that runs the test suite is 32-bit (so it should, in theory, skip this, but it doesn't). I disabled the check in https://github.com/pandas-dev/pandas/pull/57896/commits/d54d198a3824a6c13150a494b7f4b2af247737aa which I hope should fix it.

File: .github/workflows/emscripten.yml
Comment (mroeschke) [inline .github/workflows/emscripten.yml:18]: Could you add this job in `.github/workflows/unit-tests.yml` instead?

Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:18]: Sure!

File: pandas/util/_test_decorators.py
Comment (mroeschke) [inline pandas/util/_test_decorators.py:121]: Could you use your `WASM` variable here?

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml]: I had kept them in the same place for ease of access. I have now added a comment at the top of the job about where to locate the required versions and removed these environment variables altogether, in the same commit as above: [`294ab6e`](https://github.com/pandas-dev/pandas/pull/57896/commits/294ab6e0aaad3fb93cb0874c8927ac966ca8e29b)

File: pandas/util/_test_decorators.py
Comment (agriyakhetarpal) [reply pandas/util/_test_decorators.py]: Addressed in [`cc233e4`](https://github.com/pandas-dev/pandas/pull/57896/commits/cc233e4c88725440528c1b1aa690843c1d756508), along with a few other style changes across files (I activated the linter locally).

**Comment by agriyakhetarpal** [comment]
Pyodide had a new release recently to fix an error in the `pyodide build` command. I have just merged `main` and pushed a944f5287dce1439778f32525ebf5d973e75b817 to fix that. The next step from my end would be to fix some of the failing tests (which I did not have much progress on the last time and I don't want them to xfail them yet), but also to wait here until #53007 lands – I suppose that it would make running the tests against a built wheel a bit more difficult, but without the PR getting merged, I'm not sure how that will be navigated here. The good thing is that `pandas-tests` will be a pure Python package and will have a few APIs to collect and pass data files around, and installation for pure Python wheels is supported out-of-the-box in Pyodide.

If there's something else here that I should do in the meantime, I'm happy to do that – open to suggestions about resolving the two failing tests.

**Comment by agriyakhetarpal** [comment]
Putting up my response to the above comment that I am still interested in working on this.

**Comment by rgommers** [comment]
@agriyakhetarpal it seems reasonable to either skip the `tzlocal` test under WASM or lower the test tolerance specifically when running under WASM. If that is all that is needed to get the test suite to pass, I suggest doing that and then marking this as ready for review. It's easier for reviewers to give feedback when something works already.

**Comment by agriyakhetarpal** [comment]
Just merged the `main` branch, which re-triggered CI – would be good to see if some new tests or features have been added that will fail, before I resolve everything.

File: .github/workflows/unit-tests.yml
Comment (mroeschke) [inline .github/workflows/unit-tests.yml:382]: Curious why `pydantic` needs pinning?

File: .github/workflows/unit-tests.yml
Comment (mroeschke) [inline .github/workflows/unit-tests.yml:404]: Curious why pytest needs pinning?

File: pandas/tests/indexes/datetimes/methods/test_normalize.py
Comment (mroeschke) [inline pandas/tests/indexes/datetimes/methods/test_normalize.py:73]: Instead of `@td.skip_if_wasm`, could you use `@pytest.mark.skipif(WASM, reason=your comment here")` instead? (Make it clearer in the logs)

**Comment by mroeschke** [comment]
Thanks for sticking with this. It's looking close to merge

File: .github/workflows/unit-tests.yml
Comment (agriyakhetarpal) [reply .github/workflows/unit-tests.yml:404]: I have been using it for historical reasons (since the NumPy job earlier this year kept the pin, and no one bothered to remove it) – it used to be pinned earlier (https://github.com/pyodide/pyodide/issues/3978) but I believe the fix has been backported and is generally available with our current version (i.e., v0.25.1). Let me remove the pin in another commit.

Comment (agriyakhetarpal) [reply .github/workflows/unit-tests.yml:404]: Addressed in 7f4715f49b7011de2430f73dadb2acbd38b9421f.

Comment (agriyakhetarpal) [reply .github/workflows/unit-tests.yml:404]: Oops, I misread this. I thought your comment said `pydantic`, but this was `pytest`. I don't think it needs a pin, let me remove it. I deleted my earlier comments.

File: pandas/tests/indexes/datetimes/methods/test_normalize.py
Comment (agriyakhetarpal) [reply pandas/tests/indexes/datetimes/methods/test_normalize.py]: Addressed in 9f465280a0b91ca7a180b1b8e894d5a4b7b77c1f.

File: .github/workflows/unit-tests.yml
Comment (agriyakhetarpal) [reply .github/workflows/unit-tests.yml]: I had been using this pin for historical reasons (it was present due to the NumPy project having it, so I kept it here, and no one bothered to remove it). I removed it in 7f4715f49b7011de2430f73dadb2acbd38b9421f.

File: .github/workflows/unit-tests.yml
Comment (agriyakhetarpal) [reply .github/workflows/unit-tests.yml:404]: It could be unpinned. Back when this PR was opened, `pytest` had inadvertently pushed out a breaking change for many users which was later yanked, I think. Let me try to remove the pin in another commit.

Comment (agriyakhetarpal) [reply .github/workflows/unit-tests.yml]: Un-pinned in ab911d1b32bca21b48872024d60dae54d5844be5, and the test suite is proceeding to run the tests after their collection.

**Comment by agriyakhetarpal** [comment]
With most of the tests skipped where it is not possible or relevant to fix them, we are down to zero failures, and all other tests pass or xpass (or are xfailed for other reasons). The logs from the workflow run are here: https://github.com/pandas-dev/pandas/actions/runs/8945452670/job/24574452414. The doc build failures at the time of writing look to be unrelated to the changes made in this PR.

Some of the tests that have been skipped can be, perhaps, adapted to use a 32-bit version of them; I am happy to add a few of those where necessary. I'm marking this PR ready for review so that it can get the attention of more maintainers.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Great work here @agriyakhetarpal

**Comment by agriyakhetarpal** [comment]
Thanks for the merge, @mroeschke! I wish to ask if I can write a PR directly for pushing these wheels [nightly](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) to the `scientific-python` Anaconda.org index, or is it preferred to open an issue first and then link a PR to it?

For context, this is a part of the greater goal as highlighted in https://github.com/pandas-dev/pandas/issues/57891#issue-2192853050, in order to push towards interactive Sphinx-based documentation via JupyterLite. Doing this will be a much more trivial change in comparison; I am happy to contribute further.

**Comment by agriyakhetarpal** [comment]
An example that highlights some interactive documentation is available for the API reference for PyWavelets: https://pywavelets.readthedocs.io/en/latest/ref/dwt-discrete-wavelet-transform.html (work on the Usage Examples section is underway)

**Comment by mroeschke** [comment]
Feel free just to open a PR directly to add the wheels configurations to `.github/workflows/wheels.yml` and `pyproject.toml`

**Pull request merged**
